lib/identifiers/doi.rb

```diff
@@ -33,10 +33,10 @@ def self.extract(str)
         .to_s
         .downcase
         .scan(PATTERN)
-        .map { |doi|
-          next doi if doi =~ VALID_ENDING
+        .flat_map { |doi|
+          next [doi] if doi =~ VALID_ENDING
```

Comment:
The alternative to this would be to extract a separate `extract_one` method which handles the recursive case, e.g.

```ruby
    def self.extract(str)
      str
        .to_s
        .downcase
        .scan(PATTERN)
        .map { |doi| extract_one(doi) }
        .compact
    end

    def self.extract_one(str)
      match = str.to_s.downcase[PATTERN]
      return extract_one(match.sub(/\p{Punct}\z/, '')) if match && match !~ VALID_ENDING

      match
    end
```

It's not clear to me which is easier to understand.